### PR TITLE
SAK-34477 - Let user configure custom delimiters in FIB questions.

### PR DESCRIPTION
**Pull request by danielmerino**

### CODE REVIEW

File: samigo/samigo-api/src/java/xml/author/v1p2/fibTemplate.xml [hunk 1]
```diff
@@ -38,26 +38,30 @@
         <fieldlabel>ITEM_TAGS</fieldlabel>
         <fieldentry/>
       </qtimetadatafield>
-	  <qtimetadatafield>
+      <qtimetadatafield>
         <fieldlabel>ATTACHMENT</fieldlabel>
         <fieldentry/>
       </qtimetadatafield>
+      <qtimetadatafield>
+￼        <fieldlabel>MARKERS_PAIR</fieldlabel>
```
Comment (FerAlvarez95) [inline samigo/samigo-api/src/java/xml/author/v1p2/fibTemplate.xml:46]: Please remove the wrong symbol

Comment (danielmerino) [reply samigo/samigo-api/src/java/xml/author/v1p2/fibTemplate.xml]: Done. Sorry, they came in the patch I applied and I missed them.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 2]
```diff
@@ -1962,17 +1964,17 @@ public void populateFib(ItemDataIfc item, ItemContentsBean bean, Map<Long, Answe
...
+  private static ArrayList extractFIBFINTextArray(String alltext, String markers_pair)
   {
-    List texts = new ArrayList();
+    ArrayList texts = new ArrayList();
+    String tmp2="";
 
-    while (alltext.indexOf("{}") > -1)
-    {
-      int alltextLeftIndex = alltext.indexOf("{}");
+    while (alltext.indexOf(markers_pair) > -1) {
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:1972]: Tested on my local instance. Infinite loop
Should be the same as the original and only change the parameter

Comment (FerAlvarez95) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:1972]: It should be like this:

`  private static ArrayList extractFIBFINTextArray(String alltext, String markers_pair)
  {
	ArrayList texts = new ArrayList();

	while (alltext.indexOf(markers_pair) > -1) {
	  int alltextLeftIndex = alltext.indexOf(markers_pair);
      //int alltextRightIndex = alltext.indexOf("}");

      String tmp = alltext.substring(0, alltextLeftIndex);
      alltext = alltext.substring(alltextLeftIndex + 2);
      texts.add(tmp);
      // there are no more "{}", exit loop. 
      // why do we this check? will it ever come to here?
      if (alltextLeftIndex == -1)
      {
        break;
      }
    }
	texts.add(alltext);
    return texts;
  }`

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp [hunk 3]
```diff
@@ -114,12 +169,40 @@
...
+            <h:outputText value="#{authorMessages.fib_note_5}" />
+            <h:message for="customMarker" styleClass="validate" />
+        </div>
+    </div>
+
+    <div class="form-group row" id="customMarkerSettings" style="display: none;">
+        <h:outputLabel value="Custom Markers" styleClass="col-md-4 col-lg-2 form-control-label" />
+        <div class="col-md-6">
+            <h:inputText id="newmarkers" style="width: 50px;" value="#{itemauthor.currentItem.markersPair}"
+                required="false" styleClass="form-control">
```
Comment (FerAlvarez95) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp:191]: Adding maxlength="2"

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java (hunk 2 above)
Comment (danielmerino) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java]: Codacy doesn't allow to modify a parameter inside its function. I have fixed the bug using a temp variable. Thanks!

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp (hunk 3 above)
Comment (danielmerino) [reply samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp]: Good catch. Done, thanks.

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp [hunk 4]
```diff
@@ -114,12 +169,40 @@
...
+                <f:selectItem itemValue="true" itemLabel="#{authorMessages.fib_marker_custom}"
+                    id="custom" />
+            </h:selectOneRadio>
+            <h:outputText value="#{authorMessages.fib_note_5}" />
+            <h:message for="customMarker" styleClass="validate" />
+        </div>
+    </div>
+
+    <div class="form-group row" id="customMarkerSettings" style="display: none;">
+        <h:outputLabel value="Custom Markers" styleClass="col-md-4 col-lg-2 form-control-label" />
```
Comment (FerAlvarez95) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp:188]: Custom Markers is hardcoded

**Comment by FerAlvarez95** [comment]
Please, add spanish properties:

pool_badmarkers_error_1=Los delimitadores deben ser dos caracteres.
pool_badmarkers_error_2=Ambos delimitadores son iguales, deben ser diferentes.
pool_badmarkers_error_3=No puede utilizar los siguientes delimitadores:  " ' . , | * y espacio en blanco
fib_marker_default=Por defecto
fib_marker_custom=Personalizado
fib_marker=Delimitadores
fib_note_1=Establecer entre delimitadores "{}" las palabras que se quieran sustituir por un espacio en blanco para que el estudiante coloque una respuesta. <br />Ejemplo\: Las rosas son {rojas} y las violetas {azules} .
fib_note_4=*Los delimitadores deben ser caracteres diferentes.<br/>*Caracteres reservados " ' . , | * y espacio en blanco no deben ser utilizados.
fib_note_5=*Utiliza otros delimitadores si necesitas { o } en tu pregunta.
pool_missingBracket_error=Por favor, aseg\u00farese de que los delimitadores con la relaci\u00f3n y su respuesta contienen al menos un car\u00e1cter no blanco.

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp (hunk 4 above)
Comment (FerAlvarez95) [reply samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp:188]: I would suggest change value for a property like:

fib_marker_custom_enable=Custom Markers

and in spanish properties:

fib_marker_custom_enable=Delimitadores Personalizados

Comment (danielmerino) [reply samigo/samigo-app/src/webapp/jsf/author/item/fillInTheBlank.jsp]: Done, thanks.

**Comment by danielmerino** [comment]
@FerAlvarez95 @jesusmmp I added the spanish properties yesterday FYI.

**Comment by danielmerino** [comment]
@dnhutchins this patch is a backport of the work you did for your Sakai 11, could you review this patch if you have some time, please?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 5]
```diff
@@ -240,6 +240,15 @@ public void processAction(ActionEvent ae) throws AbortProcessingException {
     
     if(iType.equals(TypeFacade.FILL_IN_BLANK.toString())){
 	
+    	int markerError = isErrorMarkersFIB();
+    	if (markerError > 0) {
+    		err =ContextUtil.getLocalizedString("org.sakaiproject.tool.assessment.bundle.AuthorMessages", "pool_badmarkers_error_" + markerError);
+    		context.addMessage(null, new FacesMessage(err));
+    		item.setOutcome("fillInBlackItem");
+    		item.setPoolOutcome("fillInBlackItem");
```
Comment (dnhutchins) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:248]: Should this me FillInBlankItem instead?

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:248]: Existing references to fillInBlackItem:

`samigo/samigo-app/src/webapp/WEB-INF/faces-config.xml
590:      <from-outcome>fillInBlackItem</from-outcome>

samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/StartCreateItemListener.java
212:                        nextpage = "fillInBlackItem";

samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java
246:                    item.setOutcome("fillInBlackItem");
247:                    item.setPoolOutcome("fillInBlackItem");

samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemModifyListener.java
332:                        nextpage = "fillInBlackItem";`

**Comment by ottenhoff** [comment]
This seems like a great contribution.

Where would a good place for a unit test be? This seems like very delicate and important code. Getting some unit tests showing various examples would help a lot (similar to the calculated question unit tests).

**Comment by danielmerino** [comment]
@ottenhoff I'll take a look on the calculated questions unit tests and see if I can copy them, but I haven't never done unit tests before and I'm an occasional programmer. I don't know how much time I will spend or even if I will be able to finish them.

If anybody wants to add unit tests to this PR, it is really welcome.

**Comment by danielmerino** [comment]
Due to the pandemic situation affecting us I'm afraid I won't be able to work on this in two or three months.

It's sad that this pretty feature is stuck only by a missing test plan. If anybody else can do it, he is more than welcome!

**Comment by FerAlvarez95** [comment]
I tried to make unit tests, but the class that I wanted to test (ItemAddListener.java) uses ContextUtils in the methods I'm interested. I tried mocking FacesContext and mocking the ExternalContext too, but it didn't work. Some suggestions?
Thanks!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 6]
```diff
@@ -1962,17 +1964,17 @@ public void populateFib(ItemDataIfc item, ItemContentsBean bean, Map<Long, Answe
     bean.setFibArray(fibs);
   }
 
-  private static List extractFIBFINTextArray(String alltext)
+  private static ArrayList extractFIBFINTextArray(String alltext, String markers_pair)
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:1967]: why change from List to ArrayList?

**Comment by FerAlvarez95** [comment]
I already done a new PR adding tests and some little changes. I needed the code in this PR, so now this PR can be closed. The new one is https://github.com/sakaiproject/sakai/pull/8905. Thanks!

**Comment by danielmerino** [comment]
Closing this PR as the work is continued by @FerAlvarez95 in another one. Thanks, Fernando!

**Pull request closed without merging**